More Polishing

tarzzz opened the pull request.

n-riesco commented on Procfile:
```diff
@@ -1 +1 @@
-web: PLOTLY_CONNECTOR_PORT=$PORT PLOTLY_CONNECTOR_STORAGE_PATH="/app/sample-storage" node ./dist/headless-bundle.js
+web: PLOTLY_CONNECTOR_PORT=$PORT PLOTLY_CONNECTOR_STORAGE_PATH="/app/sample-storage" PLOTLY_CONNECTOR_AUTH_ENABLED=false node ./dist/headless-bundle.js
```
s/node/electron/

tarzzz replied:
@n-riesco Getting failure on deploys, when using `electron` instead of `node`:

```
2017-10-24T12:44:07.793994+00:00 app[web.1]: /app/node_modules/electron/dist/electron: error while loading shared libraries: libXss.so.1: cannot open shared object file: No such file or directory
```

Keeping `node` until we figure this out..

n-riesco replied:
Oh I see, the `COPY` [here](https://github.com/plotly/falcon-sql-client/blob/b7a257ac615c7ab3a9c79aa47b63f17e2bf3be00/Dockerfile#L42) should be move back to its original position (i.e. the `COPY` must happen after `yarn run rebuild:modules:eklectron`).

tarzzz replied:
In that case, the `scripts/install-modules.js` file is not copied onto image, and `yarn run rebuild:modules:electron` fails. Error:

```
module.js:471 throw err; ^ Error: Cannot find module '/usr/src/app/scripts/install-modules.js' 
at Function.Module._resolveFilename (module.js:469:15) at Function.Module._load 
(module.js:417:25) at Module.runMain (module.js:604:10) at run (bootstrap_node.js:389:7) at 
startup (bootstrap_node.js:149:9) at bootstrap_node.js:504:3
```

n-riesco replied:
What doesn't `scripts/install-modules.js` get copied into the ocntainer?

tarzzz replied:
We need to do it manually, which we are doing here: https://github.com/plotly/falcon-sql-client/pull/208/files#diff-3254677a7917c6c01f55212f86c57fbfR42

n-riesco replied:
I'm now confused.  This is what I'd expect should work:

`PLOTLY_CONNECTOR_AUTH_ENABLED=false electron ./dist/headless-bundle.js`

and
 
```
COPY . /usr/src/app

# Install app dependencies
RUN yarn install
RUN yarn run rebuild:modules:electron
```

and  `PLOTLY_CONNECTOR_AUTH_ENABLED=false node ./dist/headless-bundle.js` should fail.

chriddyp replied:
Right - so Heroku doesn't use the `Dockerfile`, heroku just installs the files from `yarn.lock` and then runs the `heroku-postbuild` script: https://github.com/plotly/falcon-sql-client/blob/master/package.json#L11. There are other ways that you can customize this process here: https://devcenter.heroku.com/articles/nodejs-support#customizing-the-build-process

tarzzz commented:
@chriddyp This adds the following updates:

  - Fixes quay.io builds: Which were failing because the scripts were being run before being copied onto the image.
  - Disable Auth for Heroku deploys: The default deploys onto Heroku should have the Auth-Disabled, since it is used for demos and testing .. 
  - Do not show login-page if auth-disabled in web-app. 

Please review..

tarzzz commented:
Update: Disabled SSL Certificates generation when in webapp/server mode.

n-riesco commented:
@tarzzz After a more careful reading of the first error message you posted:

```
2017-10-24T12:44:07.793994+00:00 app[web.1]: /app/node_modules/electron/dist/electron: error while loading shared libraries: libXss.so.1: cannot open shared object file: No such file or directory
```

I understand now what the problem is. Electron can't run inside the docker container because it requires an X server (even if no BrowserWindow is created).

As it stands, you're able to launch the web app using `node`, but it will fail as soon as any of the native modules is imported (e.g. when trying to use the DB2 connector).

To be able to launch the web app with `node` you need to:
- remove `yarn run rebuild:modules:electron` from the `Dockerfile`.
- change back the target in `webpack.config.headless.js` to `node`
- change `start-headless` in `package.json to use `node`

n-riesco commented:
@chriddyp I can see an `ENTRYPOINT` in the `Dockerfile` and a `Procfile`. Which one is in use?

tarzzz commented:
> As it stands, you're able to launch the web app using node, but it will fail as soon as any of the native modules is imported (e.g. when trying to use the DB2 connector).

That's not true actually, since the heroku hosted app works and connects to ibm DB2 fine:

http://plotly-db-connector.herokuapp.com/database-connector

chriddyp commented:
> > I can see an ENTRYPOINT in the Dockerfile and a Procfile. Which one is in use?

> Procfile is used for deploying to heroku, while Entrypoint/Docker Images is used for on-premise installations.

👍

n-riesco commented:
> Procfile is used for deploying to heroku, while Entrypoint/Docker Images is used for on-premise installations.

`Procfile` doesn't have any install/build instructions. What commands does Heroku use to install and build the webapp?

chriddyp commented:
Right - so Heroku doesn't use the Dockerfile, heroku just installs the files from yarn.lock and then runs the heroku-postbuild script: https://github.com/plotly/falcon-sql-client/blob/master/package.json#L11. There are other ways that you can customize this process here: https://devcenter.heroku.com/articles/nodejs-support#customizing-the-build-process

tarzzz commented:
Unrelated to the above conversation, but the commit [Do not generate SSL Certificates for server version.](https://github.com/plotly/falcon-sql-client/commit/d45c86698f731e2f0bfc56228e0b073d8e299dc4) causes one of the integration tests to fail.. I am going to remove that commit(via a rebase) to check if the tests pass.. and then re-introduce it with an updated test. Other commits should be good to review meanwhile ..

n-riesco commented:
@chriddyp so... If I've read the doc correctly, in our case Heroku runs the following:

1. `yarn install`
2. `yarn run heroku-postbuild` from `package.json
3. `yarn run build-headless && yarn run build-web` from `Procfile`

This explains why the native modules have been built againt `node`.

I still think that the om-premise installation is broken. I'm copying here again the changes that I still think are necessary:

- [ ] remove `yarn run rebuild:modules:electron` from the `Dockerfile`.
- [ ] change back the target in `webpack.config.headless.js` to `node`.
- [ ] change `start-headless` in `package.json` to use node

tarzzz commented:
@n-riesco I think you are confusing between on-premise and heroku builds. My attempt to clarify:

### On Premise:

  - Uses Dockerfile (and _not_ Procfile) to create a docker image, which we distribute to on-premise accounts.
  - Does `yarn run install`, `yarn run rebuild:modules:electron` , `yarn build .. etc.` as mentioned in Dockerfile
  - Uses Entrypoint to start the server: `yarn run start-headless` , which translate to the following from package.json:
```
    "start-headless": "cross-env NODE_ENV=production electron ./dist/headless-bundle.js",
```

### heroku builds:

  - Heroku automatically runs yarn install  --> package.json (It assumes yarn installation requirement from the presence of yarn.lock file, as far as I know).
  - Then it triggers post-build script: `yarn run heroku-postbuild` -- which translates to the following from package.json:
```
    "heroku-postbuild": "yarn run build-headless && yarn run build-web",

```
  - Finally, it starts the server from Procfile:

```
 node ./dist/headless-bundle.js
```

tarzzz commented:
Update: Discussed more with @n-riesco offline, and it seems that we need those updates to get on-premise working.. Added a commit which addresses the same..

chriddyp commented on Dockerfile:
```diff
@@ -39,13 +39,11 @@ ENV PLOTLY_CONNECTOR_DATA_FOLDER="/plotly-connector"
-COPY package.json /usr/src/app
-COPY yarn.lock /usr/src/app
 RUN yarn install
-RUN yarn run rebuild:modules:electron
```
So, the reason that we run each of these `COPY` commands is so that the output of the `yarn install` command is cached and not reused if `yarn.lock` doesn't change. For reference, see:
- http://blog.nchaulet.fr/cache-npm-install-on-docker/
- https://stackoverflow.com/a/35774741/4142536

tarzzz replied:
Oh I see.. that's new information.. 😸

chriddyp commented on app/components/Configuration.react.js:
```diff
@@ -53,6 +54,7 @@ class Configuration extends Component {
       cookie.remove('db-connector-user');
       cookie.remove('plotly-auth-token');
       cookie.remove('db-connector-auth-token');
+      cookie.remove('auth-disabled');
```
`db-connector-auth-disabled`? These types of typos are why we create constants and tests :)

chriddyp commented:
looks good to me! however, i still see an application error on heroku:
![image](https://user-images.githubusercontent.com/1280389/31976606-8364dc92-b906-11e7-86c8-23ef56006cb7.png)

tarzzz commented:
@chriddyp : 
  - Fixed the heroku and Docker builds (They are working now). The fix was to keep target as `electron`, but use `node` for running server.
  - Fixed issue reported by @cldougl which was causing default query to give error.
  - Addressed 🐐 comment.

I hijacked your `yarn upgrade` commit to make appveyor builds working here as well.. 😸 
Let me know if it looks good!! 

cc: @n-riesco ..

chriddyp commented:
> The fix was to keep target as electron, but use node for running server.

Can you explain this to me? I don't understand this.

chriddyp commented on app/components/Settings/Preview/CodeEditorField.react.js:
```diff
@@ -56,7 +56,8 @@ class CodeEditorField extends Component {
             updatePreview({
-                codeSchema: tables
+                codeSchema: tables,
+                code: value
```
I'm not sure if I understand this flow. We're already updating the code with `this.props.onChange`. I don't see where `this.props.preview.code` is being used anywhere else

tarzzz replied:
Now, we are supplying default queries via `code` prop on Editor Component initial render. `onChange` is not called if we do not edit the query, but directly hit `Run` after loading the ta..  this ensures that the `preview.code` is set to the default query (earlier it was null)..

n-riesco commented:
>> The fix was to keep target as electron, but use node for running server.
> Can you explain this to me? I don't understand why this works

@chriddyp Here's the reasoning behind the changes:
- `./dist/headless-bundle.js` imports `electron` and for that reason it requires the target to be `electron` (if we set the target to `node`, then `node ./dist/headless-bundle.js` will fail claiming `electron` isn't properly installed).
- the reason to use `node` to launch `./dist/headless-bundle.js` is that `electron ./dist/headless-bundle.js` fails inside the Docker container (it fails because no X server is installed and running inside the container).

The long term solution is to refactor the code (e.g. `backend/routes.js`)  so that `electron` isn't imported inside `./dist/headless-bundle.js`; I'm suggesting something like this in the long term:
- split code into three subfolders:
  - `main` for electron's main process (imports `electron` and creates a renderer process using `BrowserWindow`),
  - `renderer` for electron's renderer process (may import `electron` to communicate with the main process),
  - and `backend` for functionality that musn't import `electron` (so that it can safely be run by either `electron` or `node`)
- also I'd take this opportunity to move the backend bundles from the `backend` folder into the `dist` folder (having bundled files inside a source folder is very inconvenient when running `grep`).

chriddyp commented:
💃  - thanks for making all of those changes!
